### PR TITLE
feature: add function concat_ws

### PR DESCRIPTION
**Pull request by broccoliSpicy**
**What type of PR is this?**

- [ ] API-change
- [ ] BUG
- [ ] Improvement
- [ ] Documentation
- [x] Feature
- [ ] Test and CI
- [ ] Code Refactoring

**Which issue(s) this PR fixes:**

issue #2932 

**What this PR does / why we need it:**

Not Available

**Special notes for your reviewer:**

Not Available

**Additional documentation (e.g. design docs, usage docs, etc.):**

Not Available

### CODE REVIEW

File: pkg/sql/plan2/function/function.go [hunk 1]
```diff
@@ -258,6 +258,11 @@ func strictTypeCheck(args []types.T, require []types.T, _ types.T) bool {
 	return true
 }
 
+// todo(broccoli): change this to a general function
+func concat_ws_TypeCheck(args []types.T, require []types.T, _ types.T) bool {
```
Comment (m-schen) [inline pkg/sql/plan2/function/function.go:262]: you should just put it in the register map or concat_ws.go

**Review by m-schen — APPROVED** [review]
Reposition some of the code like concat_ws_TypeCheck function.
Put the number of argument types and type checking in type-check-function.
And I think there's no need to put concat_ws related function in 2 places (multi and vectorize), that's belongs to you.

others are good.
pass the bvt first.

**Comment by broccoliSpicy** [comment]
> Reposition some of the code like concat_ws_TypeCheck function. Put the number check and type check of function params in the type-check-function. And I think there's no need to put concat_ws related function in 2 places (multi and vectorize), that's belongs to you.
> 
> others are good. pass the bvt first.

I still wanna put the input check inside function implementation because I do input checks not only on input type, 
but also Scale/ScaleNull etc., I don't want to separate the whole input check logic, would make the code confusing if I do so

**Comment by m-schen** [comment]
> > Reposition some of the code like concat_ws_TypeCheck function. Put the number check and type check of function params in the type-check-function. And I think there's no need to put concat_ws related function in 2 places (multi and vectorize), that's belongs to you.
> > others are good. pass the bvt first.
> 
> I still wanna put the input check inside function implementation because I do input checks not only on input type, but also Scale/ScaleNull etc., I don't want to separate the whole input check logic, would make the code confusing if I do so

Both are available.
There are several advantages to having the parameter types judged outside the
1. It is beneficial to do implicit type conversion, because type-check-fn can directly determine whether the parameter type matches the expectation or not.
2. can rejecting some of the cases that do not conform to expectations in advance during the build planning phase but not execute the function.

it's not a big problem now.

**Pull request merged**